lib/rack/multipart/parser.rb

```diff
@@ -59,7 +59,9 @@ class Parser
       BUFSIZE = 1_048_576
       TEXT_PLAIN = "text/plain"
       TEMPFILE_FACTORY = lambda { |filename, content_type|
-        Tempfile.new(["RackMultipart", ::File.extname(filename.gsub("\0", '%00'))])
+        extension = ::File.extname(filename.gsub("\0", '%00'))[0..128]
```

Comment:
This behaviour is kind of complex, and I think we should test it in isolation.

Maybe introduce a method like `tempfile_name` which takes the filename and content type and returns a new file name?

I wonder if we should filter out anything except for ASCII characters from the extension (i.e. be more defensive).

Reply:
Looks like we don't use the filename, only the extension, and I cannot think of a reasonable extension that uses non-ASCII characters. However, that gsub is probably a bad idea, as we don't want to use `%00` for a replacement and we need to handle `.`.  Also, because this is only for extensions, I don't think we need to make the maximum length configurable.

For the limit, `[0, 129]` is faster than `[0..128]`.  And we can use `gsub!` and not `gsub` to save a string allocation.  Maybe something like:

```rb
ext = ::File.extname(filename)
ext.gsub!(/[^.a-zA-Z0-9]/, '')
ext = ext[0, 129]
```

This could result in ext being `.` in pathological cases, but `Dir::Tmpname`  handles that fine.

Regarding `# Does this need to be an array?`, yes, because that's what `Dir::Tmpname` handles for a prefix and suffix.

In terms of `self.tempfile_name`, if we are going to make this a method, I'd rather the method return the `Tempfile`, and we do: `TEMPFILE_FACTORY = method(:tempfile_factory)` instead of a separate lambda.  I'm OK with adding both `tempfile_name` and `tempfile_factory`, and having `tempfile_factory` call `tempfile_name` to get the name to use if you want to test it separately.

Reply:
I don't like the idea of adding methods for this, I think it makes the solution more complex than it needs to be. To do `TEMPFILE_FACTORY = method(:tempfile_factory)` the class methods needs to be defined before `TEMPFILE_FACTORY` is defined. IMHO makes the Ruby code looks a bit weird. Also, do we want these methods part of the interface?

Without `gsub("\0", '%00')`, with this code

> ```ruby
> ext = ::File.extname(filename)
> ext.gsub!(/[^.a-zA-Z0-9]/, '')
> ext = ext[0, 129]
> ```

this spec failed

```
  1) Error:
Rack::Multipart#test_0033_parse multipart form with a null byte in the filename:
ArgumentError: path name contains null byte
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:73:in `extname'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:73:in `tempfile_name'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:69:in `tempfile_factory'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:189:in `call'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:189:in `on_mime_head'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:346:in `handle_mime_head'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:262:in `block in parse'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:254:in `loop'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:254:in `parse'
    /Users/dentarg/src/rack/lib/rack/multipart/parser.rb:136:in `parse'
    /Users/dentarg/src/rack/lib/rack/multipart.rb:37:in `parse_multipart'
    test/spec_multipart.rb:389:in `block (2 levels) in <main>'
```

Reply:
I'm fine with not extracting methods provided we can test it easily.

Reply:
I'm basically okay with it, I already approved this PR. But I did wonder if we should be a bit more strict about extensions allowed, e.g. we allow non-printable characters as extensions?

Reply:
Great. I think we should try to do one thing at once. Seems like non-printable characters hasn't been a problem so far? It could be a separate contribution in the future.

Reply:
If @jeremyevans is fine with this PR, I am too.

I'm not really sure under what circumstances non-printable characters could be an issue but it seems like a super weird thing to try and write to the file-system.